src/discordtipbot/cache.cr

```diff
@@ -0,0 +1,84 @@
+class Cache
+  # Container for a snowflake and the last time it was seen
+  record Entry, id : UInt64, seen : Time do
+    def ==(other)
+      id == other.id
+    end
+
+    def <(other)
+      seen < other.seen
+    end
+
+    def >(other)
+      seen > other.seen
+    end
+
+    def hash
+      id.hash
+    end
+
+    def older_than?(age : Time::Span)
+      Time.now > seen + age
+    end
+  end
+
+  @cache = {} of UInt64 => Set(Entry)
+
+  def initialize(@ttl : Time::Span)
+  end
+
+  # Marks the `user_id` as being observed in `channel_id` at the given `time`
+  def touch(channel_id : UInt64, user_id : UInt64, time = Time.now)
+    entry = Entry.new(user_id, time)
+    set = @cache[channel_id] ||= Set(Entry).new
+    set.delete(entry) if set.includes?(entry)
+    set.add(entry)
+    @cache[channel_id] = set
+  end
+
+  # Only marks `user_id` as being observed in `channel_id` if `time` is newer then latest entry
+  def add_if_youngest(channel_id : UInt64, user_id : UInt64, time : Time)
+    entry = Entry.new(user_id, time)
+    set = @cache[channel_id] ||= Set(Entry).new
+    set.delete(entry) if set.any? { |x| x == entry && entry > x }
+    set.add(entry)
+    @cache[channel_id] = set
+  end
+
+  # Returns the set of observed entries in the given `channel_id` that have not expired.
+  def resolve(channel_id : UInt64)
+    prune(channel_id)
+  end
+
+  def resolve_to_id(channel_id : UInt64)
+    users = Set(UInt64).new
+    set = resolve(channel_id)
+    return unless set
+    set.map do |x|
+      users << x.id
+    end
+
+    users
+  end
+
+  # Clears all cached sets of expired entries
+  # or clears cached sets of expired entries of a certain channel
+  def prune(channel_id : UInt64? = nil)
+    if channel_id
+      set = prune_set(@cache[channel_id])
+      @cache[channel_id] = set # if set
+    else
+      @cache.each do |channel_id, set|
+        set = prune_set(@cache[channel_id])
+        @cache[channel_id] = set if set
+      end
+    end
+  end
+
+  private def prune_set(set : Set(Entry))
+    set.each do |entry|
+      set.delete(entry) if entry.older_than?(@ttl)
+    end
+    set
```

Comment:
Why is the extra `set` required here?

Reply:
https://carc.in/#/r/3ivl

Reply:
What I expected 😄 
(Somehow my carc.in testing showed different results though since I was being stupid once again)

Reply:
This is changed on the next crystal release I think, so that it will return self.